core/src/actors/blocks_manager/mod.rs

```diff
@@ -84,4 +112,161 @@ impl BlocksManager {
             })
             .wait(ctx);
     }
+
+    fn process_new_block(&mut self, block: Block) -> Result<Hash, BlocksManagerError> {
+        // Calculate the hash of the block
+        let hash = calculate_sha256(&block.to_bytes()?);
+
+        // Check if we already have a block with that hash
+        if let Some(_block) = self.blocks.get(&hash) {
+            return Err(BlocksManagerError::BlockAlreadyExists);
```

Comment:
Please avoid having multiple exist points for a function unless strictly necessary.